File: Dai-Hentai/ViewControllers/LoginViewController/LoginViewController.m
//
//  LoginViewController.m
//  Dai-Hentai
//
//  Created by DaidoujiChen on 2018/1/4.
//  Copyright © 2018年 DaidoujiChen. All rights reserved.
//

#import "LoginViewController.h"
#import "Dai_Hentai-Swift.h"

@interface LoginWebViewController ()

@property (nonatomic, strong) NSTimer *checkTimer;
@property (nonatomic, copy) void (^completion)(void);

@end

@implementation LoginWebViewController

#pragma mark - Private Instance Method

#pragma mark * Check Cookies

- (void)checkingLoop {
    if ([ExCookie isExist]) {
        if (self.completion) {
            self.completion();
        }
        [self cancelAction];
    }
}

#pragma mark * Init

- (void)cancelAction {
    __weak LoginWebViewController *weakSelf = self;
    [self dismissViewControllerAnimated:YES completion: ^{
        if (!weakSelf) {
            return;
        }
        __strong LoginWebViewController *strongSelf = weakSelf;
        [strongSelf.checkTimer invalidate];
    }];
}

- (void)initValues {
    UIBarButtonItem *cancelButton = [[UIBarButtonItem alloc] initWithBarButtonSystemItem:UIBarButtonSystemItemCancel target:self action:@selector(cancelAction)];
    self.navigationItem.rightBarButtonItem = cancelButton;
    
    // 開一個登入網頁
    UIWebView *loginWebView = [[UIWebView alloc] initWithFrame:self.view.bounds];
    [loginWebView loadRequest:[NSURLRequest requestWithURL:[NSURL URLWithString:@"https://forums.e-hentai.org/index.php?act=Login&CODE=01"]]];
    [self.view addSubview:loginWebView];
    
    // 用一個 timer 等到有正確的 cookie
    self.checkTimer = [NSTimer scheduledTimerWithTimeInterval:1.0f target:self selector:@selector(checkingLoop) userInfo:nil repeats:YES];
}

#pragma mark - Life Cycle

- (instancetype)initWithCompletion:(void (^)(void))completion {
    self = [super init];
    if (self) {
        self.completion = completion;
    }
    return self;
}

- (void)viewDidLoad {
    [super viewDidLoad];
    [self initValues];
}

@end

File: Dai-Hentai/ViewControllers/LoginViewController/LoginViewController.swift
//
//  LoginViewController.swift
//  Dai-Hentai
//
//  Created by DaidoujiChen on 2018/11/20.
//  Copyright © 2018 DaidoujiChen. All rights reserved.
//

import Foundation
import UIKit
import WebKit

class LoginViewController: UIViewController {
    
    // MARK: - Property
    
    private var completion: () -> ()
    private var checkTimer: Timer?
    
    // MARK: - Life Cycle
    
    @objc init(completion: @escaping () -> ()) {
        self.completion = completion
        super.init(nibName: nil, bundle: nil)
    }
    
    required init?(coder aDecoder: NSCoder) {
        fatalError("init(coder:) has not been implemented")
    }
    
    override func viewDidLoad() {
        super.viewDidLoad()
        initValues()
    }
    
    // MARK: - Private Function
    
    private func initValues() {
        let cancelButton = UIBarButtonItem(barButtonSystemItem: .cancel, target: self, action: #selector(cancelAction))
        navigationItem.rightBarButtonItem = cancelButton
        
        // 開一個登入網頁
        if let url = URL(string: "https://forums.e-hentai.org/index.php?act=Login&CODE=01") {
            let request = URLRequest(url: url)
            let webView = WKWebView(frame: view.bounds)
            webView.load(request)
            view.addSubview(webView)
        }
        
        // 用一個 timer 等到有正確的 cookie
        checkTimer = Timer.scheduledTimer(timeInterval: 1.0, target: self, selector: #selector(checkingLoop), userInfo: nil, repeats: true)
    }
    
    @objc private func cancelAction() {
        dismiss(animated: true) { [weak self] in
            guard let self = self else {
                return
            }
            self.checkTimer?.invalidate()
        }
    }
    
    @objc private func checkingLoop() {
        if ExCookie.isExist() {
            completion()
            cancelAction()
        }
    }
    
}
